io/disk/dlpar_vscsi.py

```diff
@@ -0,0 +1,155 @@
+        for line in output.decode("utf-8").splitlines():
+            if value in line:
+                if "dir" in line:
+                    self.disk_dir = line.split()[-1]
+        cmd = r"cat %s/inquiry" % self.disk_dir
+        output = process.system_output(cmd, shell=True)
+        self.hdisk_name = output.split()[2].strip(b'0001').decode("utf-8")
+        self.log.info(self.hdisk_name)
+        cmd = "lsmap -all|grep -p %s" % self.hdisk_name
+        output = self.run_command(cmd)
+        for line in output.splitlines():
+            if "VTD" in line:
+                self.vscsi = line.split()[-1]
+        self.log.info(self.vscsi)
+
+    def login(self, ipaddr, username, password):
+        '''
+        SSH Login method for remote server
+        '''
+        pxh = pxssh.pxssh(encoding='utf-8')
+        # Work-around for old pxssh not having options= parameter
+        pxh.SSH_OPTS = "%s  -o 'StrictHostKeyChecking=no'" % pxh.SSH_OPTS
+        pxh.SSH_OPTS = "%s  -o 'UserKnownHostsFile /dev/null' " % pxh.SSH_OPTS
+        pxh.force_password = True
+
+        pxh.login(ipaddr, username, password)
+        pxh.sendline()
+        pxh.prompt(timeout=60)
+        pxh.sendline('exec bash --norc --noprofile')
+        pxh.prompt(timeout=60)
+        # Ubuntu likes to be "helpful" and alias grep to
+        # include color, which isn't helpful at all. So let's
+        # go back to absolutely no messing around with the shell
+        pxh.set_unique_prompt()
+        pxh.prompt(timeout=60)
+        self.pxssh = pxh
+
+    def run_command(self, command, timeout=300):
+        '''
+        SSH Run command method for running commands on remote server
+        '''
+        self.log.info("Running the command on peer lpar: %s", command)
+        if not hasattr(self, 'pxssh'):
+            self.fail("SSH Console setup is not yet done")
+        con = self.pxssh
+        con.sendline(command)
+        con.expect("\n")  # from us
+        con.expect(con.PROMPT, timeout=timeout)
+        output = "".join(con.before)
+        con.sendline("echo $?")
+        con.prompt(timeout)
+        exitcode = int(''.join(con.before.splitlines()[1:]))
+        if exitcode != 0:
+            raise CommandFailed(command, output, exitcode)
+        return output
+
+    def test_dlpar(self):
+        '''
+        vscsi dlpar test
+        '''
+        cmd = "rmvdev -vdev %s" % self.hdisk_name
+        try:
+            output = self.run_command(cmd)
+            self.log.info(output)
+        except CommandFailed as cf:
+            self.fail("failed dlpar operation:remove disk, %s" % cf)
+        cmd = "mkvdev -vdev %s -vadapter %s -dev %s" % (self.hdisk_name,
+                                                        self.host, self.vscsi)
+        try:
+            output = self.run_command(cmd)
+            self.log.info(output)
+        except CommandFailed as cf:
+            self.fail("Failed dllpar operation:add disk, %s" % cf)
```

Comment:
Better to have functions to remove and add device, and call them.
Also, can have number of dlpar operations controlled by yaml parameter

Reply:
ok